File: fun/prechelt.jl
"""
    Prechelt

> Find a sequence of words such that the sequence of characters in these words exactly corresponds to the sequence of
> digits in the phone number. All possible solutions must be found and printed. The solutions are created word by word
> and if no word from the dictionary can be inserted at some point during that process, a single digit from the phone
> number can appear in the result at that position. Many phone numbers have no solution at all. Here is an example of
> the program output for the phone number “3586-75,” when the dictionary contained the words "Dali," "um," "Sao," "da,"
> "Pik," and 73,108 others:
>
>     3586-75: Dali um
>     3586-75: Sao 6 um
>     3586-75: da Pik 5
> 
> A list of partial solutions needs to be maintained by the program while processing each number, and the dictionary
> must be embedded in a supporting data structure (such as a 10-ary digit tree) for efficient access.
>
>  -- <cite>Prechelt, Lutz. (1999). Comparing Java vs. C/C++ Efficiency Differences to Interpersonal Differences. Communications of the ACM. 42. 10.1145/317665.317683.</quote>

This implementation uses the [Aho-Corasick algorithm](https://en.wikipedia.org/wiki/Aho%E2%80%93Corasick_algorithm)
with two small specializations

    * words are mapped to digits as they are inserted in the Aho-Corasick data structure
    * the data structure uses fixed length arrays to minimize memory allocations

```jldoctest
julia> replaceall("358", ["Sao", "dal"])
1-element Vector{String}:
 "dal"

```
"""

const DIGITS = Vector{Int}(undef, 128)
DIGITS[codeunits("0123456789")] = 1:10
DIGITS[codeunits("ejnqrwxdsyftamcivbkulopghz")] = DIGITS[codeunits("01112223334455666777888999")]
DIGITS[codeunits("EJNQRWXDSYFTAMCIVBKULOPGHZ")] = DIGITS[codeunits("01112223334455666777888999")]

mutable struct Node
    depth::Int
    keys::Vector{String}
    next::Vector{Union{Missing, Node}}
    fail::Node

    function Node(depth::Int=0)
        node = new()
        node.depth = depth
        node.keys = String[]
        node.next = Vector{Union{Missing, Node}}(missing, 10)
        node
    end
end

digits(s::String) = DIGITS[codeunits(s)]

function insert!(node::Node, word::String)
    for digit ∈ digits(word)
        next = node.next[digit]
        if ismissing(node.next[digit])
            node.next[digit] = next = Node(node.depth + 1)
        end
        node = next
    end
    push!(node.keys, word)
end

function add_fail_transition!(node::Node)
    fail = node.fail
    for (index, next) ∈ enumerate(node.next)
        ismissing(next) && continue
        next.fail = coalesce(fail.next[index], fail)
    end
    for next ∈ skipmissing(node.next)
        add_fail_transition!(next)
    end
end

function search(node::Node, index::Int)
    coalesce(node.next[index], node.depth == 0 ? node : search(node.fail, index))
end

function aho_corasick(words::Union{Base.EachLine,Vector{String}})
    node = Node()
    node.fail = node
    for word ∈ words
        insert!(node, word)
    end
    for next ∈ skipmissing(node.next)
        next.fail = node
    end
    for next ∈ skipmissing(node.next)
        add_fail_transition!(next)
    end
    node
end

crossjoin(heads::Vector{String}, tails::Vector{String}) = reduce(tails; init=[]) do partial, tail
    [partial; heads .* tail]
end

function replaceall(number::String, node::Node)
    """
    ```jldoctest
    julia> replaceall("358675", ["Dali", "um", "Sao", "da", "Pik"])
    3-element Vector{String}:
     "Dalium"
     "Sao6um"
     "daPik5"
    
    ```
    """    
    heads = Vector{Union{Missing,Vector{String}}}(missing, length(number))
    for (k, digit) ∈ enumerate(digits(number))
        node = next = search(node, digit)
        while length(next.keys) > 0
            rest = next.keys
            j = k - node.depth
            i = j - 1

            heads[k] = if j <= 1
                rest
            elseif j == 2
                crossjoin(heads[j], rest)
            else
                [crossjoin(heads[j], rest); crossjoin(heads[i], number[i:i] .* rest)]
            end
            next = next.fail
        end
    end
    filter(s -> count(r"\d", s) <= 1, [heads[end]; heads[end-1] .* number[end:end]])
end

replaceall(number::String, words::Vector{String}) = replaceall(number, aho_corasick(words))

File: fun/prechelt.py
from typing import Iterator

MAP = str.maketrans(
    "EJNQRWXDSYFTAMCIVBKULOPGHZejnqrwxdsyftamcivbkulopghz",
    "0111222333445566677788899901112223334455666777888999",
)
FAIL = 10
WORD = FAIL + 1


def count_digits(s: str, start: int = 0, sum=sum, map=map, isdigit=str.isdigit) -> int:
    return sum(map(isdigit, s), start)


def overlapping(haystack: str, needle: str, n: int) -> Iterator[str]:
    """Generate all overlapping replacements of needle in haystack

    >>> list(overlapping("1233345", "dy", 2))
    ['12dy345', '123dy45']
    """
    pattern, i = needle.translate(MAP), -1
    while True:
        i = haystack.find(pattern, i + 1)
        if i < 0:
            break
        yield haystack[:i] + needle + haystack[i + n:]


def replaceall(haystack: str, needle: str, *needles: str) -> Iterator[str]:
    """
    >>> list(replaceall("358675", "Dali", "um", "Sao", "da", "Pik"))
    ['daPik5', 'Sao6um', 'Dalium']
    """
    n = len(needle)
    if 0 <= count_digits(haystack, -n) <= 1:
        yield from overlapping(haystack, needle, n)
    elif needles:
        yield from replaceall(haystack, *needles)
        for replacement in overlapping(haystack, needle, n):
            yield from replaceall(replacement, *needles)


def aho_corasick(*words):
    """https://en.wikipedia.org/wiki/Aho%E2%80%93Corasick_algorithm"""
    root = [None] * WORD
    for word in words:
        node = root
        for digit in map(int, word.translate(MAP)):
            next = node[digit]
            if next is None:
                node[digit] = next = [None] * WORD  # pyright: ignore[reportArgumentType, reportCallIssue]
            node = next
        node.append(word)

    root[FAIL] = root  # pyright: ignore[reportArgumentType, reportCallIssue]
    stack = []
    append = stack.append
    for node in root[:FAIL]:
        if node:
            node[FAIL] = root
            append(node)

    pop = stack.pop
    while stack:
        curr = pop()
        fail = curr[FAIL]
        for digit, node in enumerate(curr[:FAIL]):
            if node:
                node[FAIL] = fail[digit] or fail
                append(node)
    return root


def replaceal1(haystack: str, *needles: str) -> Iterator[str]:
    """
    >>> list(replaceal1("358675", "Dali", "um", "Sao", "da", "Pik"))
    ['Dalium', 'Sao5um', 'daPik5']
    """
    root = aho_corasick(*needles)
    seen = [[""]]
    append = seen.append
    node = root
    for j, num in enumerate(haystack, 1):
        digit = int(num)
        while True:
            temp = node[digit]  # pyright: ignore[reportOptionalSubscript]
            if temp:
                node = temp
                break
            if node is root:
                break
            node = node[FAIL]  # pyright: ignore[reportOptionalSubscript]
        this: list[str] = []
        extend = this.extend
        for suf in node[WORD:]:  # pyright: ignore[reportOptionalSubscript]
            i = j - len(suf)  # pyright: ignore[reportArgumentType]
            extend(pre + suf for pre in seen[i])  # pyright: ignore[reportOperatorIssue]
            if i == 0:
                continue
            suf = num + suf  # pyright: ignore[reportOperatorIssue]
            extend(pre + suf for pre in seen[i - 1] if not count_digits(pre))
        append(this)

    yield from this  # pyright: ignore[reportGeneralTypeIssues, reportPossiblyUnboundVariable]
    yield from (pre + num for pre in seen[-2] if not count_digits(pre))  # pyright: ignore[reportPossiblyUnboundVariable]
